platforms/core-configuration/configuration-cache-base/src/main/kotlin/org/gradle/internal/cc/base/services/ConfigurationCacheEnvironmentChangeTracker.kt

```diff
@@ -64,6 +65,8 @@ class ConfigurationCacheEnvironmentChangeTracker(private val problemFactory: Pro
 
     fun systemPropertiesCleared() = mode.toTrackingMode().systemPropertiesCleared()
 
+    override fun <T : Any> withTrackingSystemPropertyChanges(action: Supplier<out T>): T = mode.maybeTracking().withTrackingSystemPropertyChanges(action)
```

Comment:
> 💭 We need to consider what to do when project configuration happens in parallel with IP

I think this is going to be as racy as modifying system properties concurrently. If we're talking about non-intersecting modifications (e.g. thread A writes property `a` inside the block and thread B writes property `b` through some other mean), this is going to be fine (we either see both writes when reconciling, or just `a`, but we won't touch `b`).

It gets hairier when there is concurrency, e.g. it might be possible to have this sequence:
1. thread A writes `property = a` under a block
2. A gets a snapshot
3. thread B writes `property = b`
4. A overwrites `SystemPropertyMutate` stored by B with its own.

Then the cached run is going to see a different property value (`a`) compared to the priming build which sees `b`.

I've tried to solve it, but the solution quickly becomes rather ugly, and I'm not even there yet. At this point I believe we should rethink the whole interception approach to make it truly thread-safe. And of course, attributing an input to a correct project may become a tricky quest too.